### PR TITLE
Eliminate the old "lib" module

### PR DESCRIPTION
**Pull request by richcarl**
The 'lib' module is a midden of miscellaneous stuff, inviting bad programming habits, and uses an extremely unspecific module name. This patch moves the still-used functionality to better locations (in a few cases duplicating some trivial code). The module was already documented as possibly disappearing without warning in a future release.

### CODE REVIEW

File: lib/dialyzer/test/options1_SUITE_data/src/compiler/compile.erl [hunk 1]
```diff
@@ -228,11 +228,15 @@ os_process_size() ->
     case os:type() of
 	{unix, sunos} ->
 	    Size = os:cmd("ps -o vsz -p " ++ os:getpid() ++ " | tail -1"),
-	    list_to_integer(lib:nonl(Size));
+	    list_to_integer(nonl(Size));
 	_ ->
 	    0
     end.
 
+nonl([10]) -> [];
```
Comment (fenollp) [inline lib/dialyzer/test/options1_SUITE_data/src/compiler/compile.erl]: How about `"\n"`?

Comment (fenollp) [reply lib/dialyzer/test/options1_SUITE_data/src/compiler/compile.erl]: Or $\n to keep the square brackets.
Actually I’d suggest lists:droplast(Size). Surely that OS call always returns a newline.

File: lib/common_test/src/ct.erl [hunk 2]
```diff
@@ -975,7 +976,20 @@ make_priv_dir() ->
...
+%%% @doc Return the command used to start (this) erlang
+
+-spec get_progname() -> string().
+
+get_progname() ->
+    case init:get_argument(progname) of
+	{ok, [[Prog]]} ->
+	    Prog;
+	_Other ->
+	    "no_prog_name"
```
Comment (fenollp) [inline lib/common_test/src/ct.erl]: Would an empty string be better here? That string is never matched against even in atom form in OTP.

Comment (richcarl) [reply lib/common_test/src/ct.erl]: No, the result is used as the base for several command lines passed to os:cmd(), and if empty, the options get interpreted as commands - better with a string that will probably never be a command, so you'll trigger some kind of error.

File: lib/dialyzer/test/options1_SUITE_data/src/compiler/compile.erl (hunk 1 above)
Comment (richcarl) [reply lib/dialyzer/test/options1_SUITE_data/src/compiler/compile.erl]: I don't like to move code and do changes at the same time. But I could add a commit that cleans up this ugly stuff. $\n is the right way.

**Comment by leoliu** [comment]
Does `nonl/1` work on windows?

**Comment by leoliu** [comment]
@richcarl thanks.

**Comment by uabboli** [comment]
Thanks! This pull request is approved by the OTP technical board.

There are a few things to fix. This is what I've found so far:

Remove specs_lib.xml from lib/stdlib/doc/src/specs.xml.

Add erl_error to the module list in lib/stdlib/src/Makefile.

Add erl_error.beam to the primary_bootstrap_compiler target in lib/stdlib/src/Makefile. And add bootstrap/lib/stdlib/ebin/erl_error.beam.

Replacing append with concat in lib/stdlib/src/slave.erl does not work since there are atoms in the argument list.

If we some day want to document erl_error, it could be a problem that erl_error(3) already exists in the erl_interface application. But we think we can handle that by adding a prefix or something to the stuff in erl_interface.

**Comment by uabboli** [comment]
Thanks. Almost there... Since eval_str is no longer exported, Dialyzer complains:

mod_esi.erl:612: The variable Bin can never match since previous clauses completely covered the type string()

**Comment by uabboli** [comment]
I noticed now that you've resolved a conflict. We don't merge master into branches, so could you rebase your branch upon the latest master instead? Thanks.

**Comment by richcarl** [comment]
Sure. I wanted to make sure everything was ready before I rebased and made the review harder.

**Comment by uabboli** [comment]
Hi again. It seems that another rebase is needed to include the new Logger functionality in proc_lib.

**Comment by richcarl** [comment]
Rebased. Don't forget to update bootstrap.

**Comment by uabboli** [comment]
I've rebuilt the bootstrap files (including erl_error.beam), and added a branch to our daily builds.

**Comment by uabboli** [comment]
The daily tests have pinpointed a few minor issues which I've fixed.
See the diffs below.

`ts_erl_config` calls `ct:get_progname()`. During cross compilation
the `ct` module is not readily available. We could add the `ct` module
to the bootstrap compiler, but we don't want to do that. Or we could
eliminate the call to `ct:get_progname()` as in the `slave` module. I
assume you've already pondered, but would it be possible to add
get_progname to a module already included in `/bootstrap`? Then
`slave` could use that function as well.

I noticed that `ets` has implemented `nonl` like this:

    nonl(S) -> string:trim(S, trailing, "$\n").

Even better is `string:chomp/1` (works well on Windows).

    ------------ lib/dialyzer/test/r9c_SUITE_data/src/inets/mod_esi.erl ------------
    index d9d27bf..ce144e0 100644
    @@ -352,9 +352,7 @@ eval_str(Str) when is_list(Str) ->
			{error, ?result("Non-white space found after "
					"end-of-form :~ts", [Rest])}
		    end
    -    end;
    -eval_str(Bin) when is_binary(Bin) ->
    -    eval_str(binary_to_list(Bin)).
    +    end.

     all_white([$\s|T]) -> all_white(T);
     all_white([$\n|T]) -> all_white(T);


    ----- lib/stdlib/src/stdlib.app.src ------------------------
    index e698ca0..0687a6b 100644
    @@ -2,7 +2,7 @@
     %% 
     %% %CopyrightBegin%
     %%
    -%% Copyright Ericsson AB 1996-2017. All Rights Reserved.
    +%% Copyright Ericsson AB 1996-2018. All Rights Reserved.
     %%
     %% Licensed under the Apache License, Version 2.0 (the "License");
     %% you may not use this file except in compliance with the License.
    @@ -43,6 +43,7 @@
		  erl_anno,
		 erl_bits,
		 erl_compile,
    +             erl_error,
		 erl_eval,
		  erl_expand_records,
		 erl_internal,

    ------------ lib/dialyzer/test/options1_SUITE_data/results/compiler ------------
    index cbb5115..e1dc038 100644
    @@ -28,7 +28,7 @@ cerl_inline.erl:2750: The pattern <{[], L, D}, Vs> can never match the type <[1.
     cerl_inline.erl:2752: The pattern <{[], _L, D}, Vs> can never match the type <[1..255,...],[any()]>
     cerl_inline.erl:2754: The pattern <{F, L, D}, Vs> can never match the type <[1..255,...],[any()]>
     cerl_inline.erl:2756: The pattern <{F, _L, D}, Vs> can never match the type <[1..255,...],[any()]>
    -compile.erl:788: The pattern {'error', Es} can never match the type {'ok',<<_:64,_:_*8>>}
    +compile.erl:792: The pattern {'error', Es} can never match the type {'ok',<<_:64,_:_*8>>}
     core_lint.erl:473: The pattern <{'c_atom', _, 'all'}, 'binary', _Def, St> can never match the type <_,#c_nil{} | {'c_atom' | 'c_char' | 'c_float' | 'c_int' | 'c_string' | 'c_tuple',_,_} | #c_cons{hd::#c_nil{} | {'c_atom' | 'c_char' | 'c_float' | 'c_int' | 'c_string' | 'c_tuple',_,_} | #c_cons{hd::{_,_} | {_,_,_} | {_,_,_,_},tl::{_,_} | {_,_,_} | {_,_,_,_}},tl::#c_nil{} | {'c_atom' | 'c_char' | 'c_float' | 'c_int' | 'c_string' | 'c_tuple',_,_} | #c_cons{hd::{_,_} | {_,_,_} | {_,_,_,_},tl::{_,_} | {_,_,_} | {_,_,_,_}}},[any()],_>
     core_lint.erl:505: The pattern <_Req, 'unknown', St> can never match the type <non_neg_integer(),non_neg_integer(),_>
     sys_pre_expand.erl:625: Call to missing or unexported function erlang:hash/2

**Comment by richcarl** [comment]
I don't think that the get_progname is sufficiently general and useful to be a library function except in ct. Its default of "no_prog_name" is arbitrary and should not be allowed to spread further. I pushed a commit to eliminate the dependency from ts_erl_config.

**Comment by richcarl** [comment]
Added a separate commit with your fixes.

**Comment by richcarl** [comment]
Yay!

**Pull request merged**